### PR TITLE
Consider all non-interactive terminals to be bots

### PR DESCRIPTION
**Pull request by tvolkert**
## Description

Flutter is seeing spikes in its flutter_tools usage reporting that appear to be bots (e.g. they come disproportionally from one specific area, or they create without ever running).  This change attempts to better detect bots by checking for whether the invocation has an interactive terminal or not.

## Tests

I added the following tests:

* A test that verifies that the `hasTerminal` check is factored into the `isRunningOnBot` check.

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by goderbauer** [comment]
Does this change how we count interactions with flutter_tool via IDEs at all? I'd assume they run it without having a proper terminal attached to it?

**Comment by tvolkert** [comment]
> Does this change how we count interactions with flutter_tool via IDEs at all? I'd assume they run it without having a proper terminal attached to it?

Good question.  @devoncarew do the IDEs set any environment variables to tell the tool that they're _not_ bots?  If they set the environment variable `BOT=false`, then we'll assume the environment variable knows best and not bother checking for other bot signals.

**Comment by devoncarew** [comment]
> Good question. @devoncarew do the IDEs set any environment variables to tell the tool that they're _not_ bots? If they set the environment variable `BOT=false`, then we'll assume the environment variable knows best and not bother checking for other bot signals.

They set the `FLUTTER_HOST` env variable to the IDE name - that would be a good env variable to key off of to ensure that it's not a bot.

File: packages/flutter_tools/lib/src/base/utils.dart [hunk 1]
```diff
@@ -24,6 +24,9 @@ class BotDetector {
     return platform.environment['BOT'] != 'false'
        && (platform.environment['BOT'] == 'true'
 
+        // Non-interactive terminals are assumed to be bots.
+        || !io.stdout.hasTerminal
```
Comment (dnfield) [inline packages/flutter_tools/lib/src/base/utils.dart]: Per Devon's comment, it seems like this should be `|| (!io.stdout.hasTerminal && platform.environment['FLUTTER_HOST'] == null)`

Comment (tvolkert) [reply packages/flutter_tools/lib/src/base/utils.dart]: I updated the code - ptal

File: packages/flutter_tools/lib/src/base/utils.dart [hunk 2]
```diff
@@ -43,7 +55,7 @@ class BotDetector {
 
         // Properties on Flutter's Chrome Infra bots.
         || platform.environment['CHROME_HEADLESS'] == '1'
-        || platform.environment.containsKey('BUILDBOT_BUILDERNAME'));
+        || platform.environment.containsKey('BUILDBOT_BUILDERNAME');
```
Comment (dnfield) [inline packages/flutter_tools/lib/src/base/utils.dart]: While we're here, can we add `|| platform.environment.containsKey('SWARMING_TASK_ID')`?  

I strongly suspect `BUILDBOT_BUILDERNAME` was used previously on Chrome infra, but it's definitely not getting set on LUCI today.  AFAICT we're not excluding LUCI runs from metrics reports.

Comment (dnfield) [reply packages/flutter_tools/lib/src/base/utils.dart]: (It should be safe to leave the current check in place though)

Comment (tvolkert) [reply packages/flutter_tools/lib/src/base/utils.dart]: Done.

**Review by dnfield — APPROVED** [review]
LGTM

**Pull request merged**